### PR TITLE
Gateway read and write on Database

### PR DESCRIPTION
**Pull request by pikonha**
- **feat: server options pattern**
- **feat: add balance and address handlers**
- **feat: mongodb repo with models**
- **feat: removed signed balance support**
- **feat: add l2 repo structure**
- **feat: replace prisma by typeorm**
- **feat: add contenthash functions to abi**
- **chore: debug tests from root**
- **chore: debug gateway test**
- **chore: sharing tsconfig and babel**
- **chore: client test using vitest**
- **fix: datasource initialize on tests**
- **fix: tests running with vitest**
- **chore: tsconfig as shared package**
- **test: setText**
- **test: readText through domain relation**
- **chore: remove ttl and add setAddr**
- **test: read/write address and text**
- **test: read/write contenthash**

# Gateway read and write on Database

## All Submissions:

* [X] Have you followed the guidelines in our Contributing document?
* [X] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

## Description
It implements the handlers to read and write data into a database through the Gateway (relying on [CCIP-Read-Server](https://github.com/smartcontractkit/ccip-read)). The structure was made modular so the gateway doesn't need to implement all the functions implemented by the PublicResolver.

TODO after this gets merged:
* #38 Support function overload, [see issue](https://github.com/smartcontractkit/ccip-read/issues/32)
* #39 Implement gateway response signing to ensure authenticity 
* #40 Implement gateway address recovering from signed data to ensure ownership

## Related Issue
#38 
#39
#40

## Changes
- [X] New feature implementation
- [ ] Bug fix
- [ ] Code refactoring
- [ ] Documentation update
- [ ] Other (please specify)

### Changes to Core Features:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [X] Have you written new tests for your core changes, as applicable?
* [X] Have you successfully run tests with your changes locally?

### CODE REVIEW

File: packages/gateway/src/entities/domain.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import { Entity, PrimaryColumn, Column, OneToMany } from 'typeorm'
+import { Address } from './address'
+import { Text } from './text'
+
+@Entity()
+export class Domain {
+  @PrimaryColumn()
+  namehash: string
```
Comment (alextnetto) [inline packages/gateway/src/entities/domain.ts:8]: Namehash is the name of the process, not of the data itself. The [ENS documents](https://docs.ens.domains/terminology#node) clarify that.

File: packages/gateway/package.json [hunk 2]
```diff
@@ -1,23 +1,35 @@
...
-  "name": "gateway",
+  "name": "@blockful/gateway",
   "version": "1.0.0",
   "description": "",
   "scripts": {
     "test": "vitest run",
     "test:watch": "vitest",
-    "start": "ts-node ./src/index.ts",
-    "lint": "eslint . --ext .ts --fix"
+    "start": "dotenvx run --env-file=../../.env -- ts-node ./src/index.ts",
```
[the remark this answers is not in the document]
Comment (pikonha) [reply packages/gateway/package.json:8]: yes, that's exactly what it does

File: packages/gateway/src/entities/address.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import { Entity, Column, PrimaryColumn, ManyToOne, JoinColumn } from 'typeorm'
```
[the remark this answers is not in the document]
Comment (pikonha) [reply packages/gateway/src/entities/address.ts:1]: done

File: packages/gateway/src/repositories/l2.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
+/* eslint-disable */
```
[the remark this answers is not in the document]
Comment (pikonha) [reply packages/gateway/src/repositories/l2.ts:1]: it is complaining about unused variables which makes the pre-commit hook to fail

File: packages/gateway/src/index.ts [hunk 5]
```diff
@@ -1,10 +1,40 @@
...
+import 'reflect-metadata'
 
-const app = server.makeApp('/')
+import { NewDataSource } from './datasources/typeorm'
+import {
+  withGetText,
+  withSetText,
+  withGetAddr,
+  withSetAddr,
+  withContentHash,
```
[the remark this answers is not in the document]
Comment (pikonha) [reply packages/gateway/src/index.ts:12]: done

File: packages/gateway/src/entities/domain.ts (hunk 1 above)
Comment (pikonha) [reply packages/gateway/src/entities/domain.ts]: done

**Review by LeonardoVieira1630 — APPROVED** [review]

**Pull request merged**